llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+import os
+from typing import Any, Dict, List, Optional, Union
+
+from composer.core import Callback, Event, State, Time
+from composer.loggers import Logger
+from composer.loggers.mosaicml_logger import (MOSAICML_PLATFORM_ENV_VAR,
+                                              RUN_NAME_ENV_VAR)
+from composer.utils import dist
+from composer.utils.misc import create_interval_scheduler
+
+from mcli import ComputeConfig, Run, RunConfig, create_run, get_run
+
+log = logging.getLogger(__name__)
+
+MAX_RUN_NAME_LENGTH = 40
+
+REQUIRED_PARAMS_FOR_EVAL = {
+    'device_eval_batch_size',
+    'icl_tasks',  # only required for eval, may not be specified in pure training
+    'max_seq_len',
+    'model',  # converted into models
+    'tokenizer',  # converted into models
+    'save_folder',  # required, but used as load_path
+}
+OPTIONAL_PARAMS_FOR_EVAL = {
+    'dist_timeout',
+    'eval_gauntlet',
+    'fsdp_config',
+    'icl_subset_num_batches',
+    'loggers',
+    'precision',
+    'python_log_level',
+    'seed',
+}
+
+
+def get_run_name(previous_run_name: str, count: int) -> str:
+    *name_without_uuid_suffix, _ = previous_run_name.split('-')
+    name_suffix = ('-'.join(name_without_uuid_suffix))
+
+    if len(name_suffix) > MAX_RUN_NAME_LENGTH:
+        log.warning(
+            f'Training run name {name_suffix} may be too long, truncating to {MAX_RUN_NAME_LENGTH} characters'
+        )
+        name_suffix = name_suffix[:MAX_RUN_NAME_LENGTH]
+
+    return f'eval{count}-{name_suffix}'
+
+
+def get_load_path(save_folder: str,
+                  save_latest_filename: Optional[str] = None) -> str:
+    # TODO: check that the prefix is remote and not a local file (not supported of course)
+
+    if not save_latest_filename:
+        rank = dist.get_global_rank()
+        save_latest_filename = f'latest-rank{rank}.pt'
+
+    return f'{save_folder}/{save_latest_filename}'
```

Comment:
So this will just use whatever the last checkpoint that has arrived in object store is? Even if it has already evaled on that checkpoint?

Reply:
yea - do you think it would make more sense to somehow pull this from state? It sort of assumes you know what you're doing when you configure the checkpointing and async eval intervals

Reply:
Unfortunately we don't really have a way to know when the checkpoint actually arrives in object store, which would be the ideal time run the asynceval callback. I previously suggested something along the lines of the eval callback keeping track of what checkpoints it has evaled and not doing the same one more than once, but that would require the eval job to be able to communicate back to the training job or something. hmmm

Reply:
Summarizing conversation with @dakinggg, and iterating over possible solutions to async eval checkpointing:
1. Callback checks in "micro intervals" if any checkpoints are ready, and if so schedules new runs
    * Would need to know when the checkpoint is ready. Reporting on when the checkpoint has finished uploading is difficult (but not impossible) because of complexity of sharded checkpoints. 
    * Completed checkpoints need to be accessible in state
2. Sidecar CPU run manages eval run creation (no async eval callback at all)
    * Need dedicated cpu instances on each cluster (one per expected hero / finetuning run)
    * Need to know when checkpoint is ready
3. Eval "wait for checkpoint" parameter and custom logic to wait until checkpoint is ready
    * GPUs sit idle until checkpoint is ready
    * Timeout needed to keep these runs not running forever. What happens on failure case?
4. Cloud support for "scheduled runs" and wait to run eval runs for fixed amount of time 
    * Checkpoint completion highly dependent on model size & network quality, "fixed" time not a great solution
5. Async callback called with latest checkpoint (pull from state in CheckpointSaver callback)
    * Intermediate checkpoints may fail if checkpoint is not ready. These eval runs can have watchdog enabled or be manually retried. It might be ok if not all intermediate checkpoint evals successfully run
    * The final checkpoint must be eval-ed. Checkpointsaver will block on fit end - Async eval should trigger after this

Going with (5), unless anyone has major feedback, ack @dakinggg @eracah @mvpatel2000

Reply:
ya, 5 is good for me for prototyping. i would guess that 2 is the long term solution though.